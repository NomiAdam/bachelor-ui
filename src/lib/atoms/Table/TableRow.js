import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { compose, tail, values } from 'ramda';
import { mapIndexed } from 'ramda-extension';
import { IoEdit, IoAndroidDelete } from 'react-icons/lib/io';
import TableCell from './TableCell';
import { lightTheme, darkTheme } from '../../constants/theme';

const StyledTableRow = styled.tr`
	background-color: transparent;
	border-bottom: 1px solid ${ lightTheme.lightGrey };
	transition: .25s all ease;
	&:hover {
		background-color: ${ ({ effect }) => (effect ? 'transparent' : lightTheme.whiteBlue) };
	}
`;

const Icon = styled.div`
    padding: 10px 15px;
    font-size: 1.5em;
    cursor: pointer; 
    color: ${ darkTheme.darkGrey };
`;

const StyledEdit = Icon.withComponent(IoEdit);
const StyledTrash = Icon.withComponent(IoAndroidDelete);

class TableRow extends PureComponent {
    state = {
        rowData: [],
    };

    componentDidMount() {
        this.handleTableCellsMap();
    }

    componentDidUpdate(prevProps) {
        const { tableRowData } = this.props;
        if (tableRowData !== prevProps.tableRowData) {
            this.handleTableCellsMap();
        }
    }

    handleTableCellsMap = () => {
        const { tableRowData, redirect, handleRedirect } = this.props;

        const mapTableCells = (value, index) => (
            <TableCell
                redirect={ redirect }
                key={ `${ index }-${ value }` }
                onClick={ handleRedirect }
            >
                {value}
            </TableCell>
        );
        const mappedTableCells = compose(tail, values, mapIndexed(mapTableCells));

        this.setState({ rowData: mappedTableCells(tableRowData) });
    };

    render() {
        const { editable, handleDelete, handleEdit } = this.props;
        const { rowData } = this.state;
        return (
            <StyledTableRow>
                {editable && (
                    <TableCell editable>
                        <StyledTrash onClick={ handleDelete } />
                        <StyledEdit onClick={ handleEdit } />
                    </TableCell>
                )}
                {rowData}
            </StyledTableRow>
        );
    }
}

TableRow.propTypes = {
    /**
     * Whether Table row should have something of Effect on hover, in our case color change
     */
    effect: PropTypes.bool,
    /**
     * Object of keys is expected
     */
    tableRowData: PropTypes.object,
    /**
     *  Whether table has extra column with edit Icons, boolean is expected
     */
    editable: PropTypes.bool,
    /**
     * Function to handle edit icon action, is present only when table is editable
     */
    handleEdit: PropTypes.func,
    /**
     * Function to handle delete icon action, is present only when table is editable
     */
    handleDelete: PropTypes.func,
    /**
     * Whether we redirect on click or not
     */
    redirect: PropTypes.bool,
    /**
     * Function to handle redirect on click, if redirect is set to false, handleRedirect is null
     */
    handleRedirect: PropTypes.any,
};

TableRow.defaultProps = {
    tableRowData: {},
    redirect: false,
    editable: false,
};

export { StyledTableRow, StyledEdit, StyledTrash };
export default TableRow;
