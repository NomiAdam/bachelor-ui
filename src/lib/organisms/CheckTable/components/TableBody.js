import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import TableRow from './TableRow';
import {range, mapObjIndexed, values, o} from 'ramda';

class ReservationTableBody extends PureComponent {
    state = {
        bodyData: [],
    };

    componentDidUpdate(prevProps) {
        const { selected } = this.props;
        if( selected !== prevProps.selected ) {
            this.handleTableBodyMap();
        }
    }

    handleTableBodyMap = () => {
        const {rowRange, handleChange, selected, isCheckbox} = this.props;

        const cellRange = range(0, rowRange);

        const mapTableRow = ({value, label, disabled}, key) => (
            <TableRow
                key={key}
                label={label}
                dayHours={cellRange}
                selected={value}
                disabled={disabled}
                handleChange={handleChange(key)}
                isCheckbox={isCheckbox}
            />
        );

        const mapRow = mapObjIndexed(mapTableRow);
        const mappedTableRows = o(values, mapRow);

        this.setState({ bodyData: mappedTableRows(selected) })
    };

    render() {
        const { bodyData } = this.state;
        return (
            <tbody>
            {bodyData}
            </tbody>
        );
    }
}

ReservationTableBody.propTypes = {
    /**
     * Number indicating how many TableCells we will have in each row
     */
    rowRange: PropTypes.number,
    /**
     * Handle onClick on TableCell, (key) => (value) => is expected
     */
    handleChange: PropTypes.func,
    /**
     * Object of current table Rows, format as below
     * { key: { label: 'label', value: [] }, ... }
     * value array is array of selected (checked) TableCells
     */
    selected: PropTypes.object,
    /**
     * Whether we use checkbox component inside table or not
     */
    isCheckbox: PropTypes.bool,
};

ReservationTableBody.defaultProps = {
    dayHours: [],
    selected: {},
    rowRange: 0,
};

export default ReservationTableBody;
