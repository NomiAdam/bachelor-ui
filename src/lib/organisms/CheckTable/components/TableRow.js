import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {TableCell} from '../../../atoms/Table/index';
import {StyledTableRow} from '../../../atoms/Table/SimpleTableRow';
import Checkbox from '../../../atoms/Checkbox/index';
import CheckCell from './CheckCell';
import {
    map, addIndex, indexOf, o, append, equals, ifElse, __, remove, always, reduce, add,
} from 'ramda';

const mapIndexed = addIndex(map);
const isSelected = equals(-1);

const computeSelectedArray = reduce(add, 0);

class TableRow extends PureComponent {
    state = {
        rowData: [],
    };

    componentDidMount() {
        this.handleDayHoursMap();
    };

    componentDidUpdate(prevProps) {
        const {selected} = this.props;
        if (computeSelectedArray(selected) !== computeSelectedArray(prevProps.selected)) {
            this.handleDayHoursMap();
        }
    };

    handleClick = (id) => (e) => {
        e.preventDefault();

        const {selected, handleChange} = this.props;

        const indexOfSelected = indexOf(__, selected);

        const appendWithSelected = always(append(id, selected));
        const removeFromSelected = always(o(remove(__, 1, selected), indexOfSelected)(id));

        const removeOrConcat = ifElse(isSelected, appendWithSelected, removeFromSelected);

        const changedSelectedState = o(removeOrConcat, indexOfSelected);

        handleChange(changedSelectedState(id));
    };

    handleDayHoursMap = () => {
        const {dayHours, isCheckbox} = this.props;
        const mapDayHour = (hour, index) => {
            return isCheckbox ?
                (
                    <TableCell key={index} onClick={this.handleClick(index)}>
                        <Checkbox checked={this.isSelected(index) && !this.isDisabled(index)}/>
                    </TableCell>
                ) : (
                    <CheckCell checked={this.isSelected(index)} disabled={this.isDisabled(index)}/>
                );
        };
        const mappedReservationCells = mapIndexed(mapDayHour);

        this.setState({rowData: mappedReservationCells(dayHours)});
    };

    isSelected = id => this.props.selected.indexOf(id) !== -1;

    isDisabled = id => this.props.disabled.indexOf(id) !== -1;

    render() {
        const {label} = this.props;
        const {rowData} = this.state;
        return (
            <StyledTableRow>
                <TableCell>
                    {label}
                </TableCell>
                {rowData}
            </StyledTableRow>
        );
    }
}

TableRow.propTypes = {
    /**
     * String representing label for each row, note that this can be day or sports center
     */
    label: PropTypes.string.isRequired,
    /**
     * Array of displayed day hours
     */
    dayHours: PropTypes.array,
    /**
     * Handle onClick on TableCell, (value) => is expected
     */
    handleChange: PropTypes.func,
    /**
     * Array of checked TableCells for current row
     */
    selected: PropTypes.array,
    /**
     * Whether we use checkbox inside component or not
     */
    isCheckbox: PropTypes.bool,
    /**
     * Array of disabled TableCells for current row
     */
    disabled: PropTypes.array,
};

TableRow.defaultProps = {
    dayHours: [],
    selected: [],
    disabled: [],
    isCheckbox: true,
};

export default TableRow;
