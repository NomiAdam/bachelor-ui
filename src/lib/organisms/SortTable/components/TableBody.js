import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {TableRow} from '../../../atoms/Table/index';
import {map, slice, o, dec, multiply, sort, prop, equals, ifElse, descend, ascend, always} from 'ramda';

const getSliceStart = o(multiply, dec);

const isAscending = equals('asc');
const sortByAscend = always(ascend);
const sortByDescend = always(descend);
const sortByOrder = ifElse(isAscending, sortByAscend, sortByDescend);

const getRowIdentification = prop('id');

class TableBody extends PureComponent {
    state = {
        columnData: [],
    };

    componentDidUpdate(prevProps) {
        const {
            columnData, order, page, rowsPerPage,
        } = this.props;
        if (
            columnData !== prevProps.columnData
            ||
            order !== prevProps.order
            ||
            page !== prevProps.page
            ||
            rowsPerPage !== prevProps.rowsPerPage
        ) {
            this.handleColumnDataMap();
        }
    };

    handleColumnDataMap = () => {
        const {
            columnData,
            rowsPerPage,
            page,
            orderBy,
            order,
            editable,
            handleDelete,
            handleRedirect,
            redirect,
            handleEdit
        } = this.props;

        const mapTableRow = (value) => (
            <TableRow
                tableRowData={value}
                key={getRowIdentification(value)}
                editable={editable}
                handleDelete={handleDelete(getRowIdentification(value))}
                handleEdit={handleEdit(getRowIdentification(value))}
                handleRedirect={redirect ? handleRedirect(getRowIdentification(value)) : null}
                redirect={redirect}
            />
        );

        const mappedTableRows = map(mapTableRow);

        const sliceStart = getSliceStart(page)(rowsPerPage);
        const sliceEnd = multiply(page, rowsPerPage);
        const sliceBody = slice(sliceStart, sliceEnd);

        const byOrder = sortByOrder(order)(prop(orderBy));

        const sortByGivenKey = sort(byOrder);
        const sortedBody = o(sortByGivenKey, sliceBody);

        const mappedBody = o(mappedTableRows, sortedBody);

        this.setState({ columnData: [] }, () => {
            this.setState({columnData: mappedBody(columnData)})
        });
    };

    render() {
        const {columnData} = this.state;
        return (
            <tbody>
            {columnData}
            </tbody>
        );
    };
}

TableBody.propTypes = {
    /**
     * Number representing current page
     */
    page: PropTypes.number,
    /**
     * Number representing how many rows per page are displayed
     */
    rowsPerPage: PropTypes.number,
    /**
     * String value, representing key, by which we will sort
     */
    orderBy: PropTypes.string,
    /**
     * String representation, either ascending, or descending
     */
    order: PropTypes.string,
    /**
     * [{id: 'id', key: 'value', anotherKey: 'val'}], note that every object must start with id key
     */
    columnData: PropTypes.array,
    /**
     * Function to handle current row delete
     */
    handleDelete: PropTypes.func,
    /**
     * Function to handle current row edit
     */
    handleEdit: PropTypes.func,
    /**
     * Boolean value, indicating whether the body is editable or not
     */
    editable: PropTypes.bool,
    /**
     * Whether we redirect on click or not
     */
    redirect: PropTypes.bool,
    /**
     * Function to handle redirect on click
     */
    handleRedirect: PropTypes.func,
};

TableBody.defaultProps = {
    columnData: [],
    editable: false,
    redirect: false,
};

export default TableBody;
