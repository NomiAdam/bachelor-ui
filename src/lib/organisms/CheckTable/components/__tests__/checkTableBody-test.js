/* eslint-disable no-undef */
import React from 'react';
import {shallow} from 'enzyme';
import TableBody from '../TableBody';
import TableRow from '../TableRow';

const mockBodyKey = 'mock1';
const mockBody = {
    [mockBodyKey]: {
        label: 'Mock 1',
        value: [],
    },
    mock2: {
        label: 'Mock 2',
        value: [],
    }
};

const mockRange = 5;
const mockChange = jest.fn();

const renderTableBody = (
    <TableBody
        rowRange={mockRange}
        selected={mockBody}
        handleChange={mockChange}
    />
);


describe('CheckTableBody', () => {
    it('should be defined', () => {
        expect(TableBody).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<TableBody />)).toMatchSnapshot();
    });
    it('should match the snapshot with data', () => {
        expect(shallow(renderTableBody)).toMatchSnapshot();
    });
    it('should be of type tbody', () => {
        expect(shallow(renderTableBody).type()).toBe('tbody');
    });
    it('should render the same amount of row as are selected props', () => {
        const wrapper = shallow(<TableBody handleChange={mockChange}/>);
        wrapper.setProps({ selected: mockBody });
        expect(wrapper.find(TableRow).length).toBe(2);
    });
});
