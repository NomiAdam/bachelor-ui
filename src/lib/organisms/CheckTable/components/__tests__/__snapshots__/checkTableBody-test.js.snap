// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckTableBody should match the snapshot 1`] = `<tbody />`;

exports[`CheckTableBody should match the snapshot with data 1`] = `<tbody />`;
